File: app/src/main/kotlin/com/taleson2wheels/app/ui/T2WApp.kt
package com.taleson2wheels.app.ui

import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.taleson2wheels.app.di.AppContainer
import com.taleson2wheels.app.ui.auth.LoginScreen
import com.taleson2wheels.app.ui.common.LoadingView
import com.taleson2wheels.app.ui.theme.T2WTheme

/**
 * Root composable + auth gate. The session lives in [AppContainer.session]; this
 * observes it to switch between the login screen and the signed-in app shell.
 * Clearing the session anywhere (logout, refresh failure) automatically returns
 * the user to login — no manual navigation required.
 */
@Composable
fun T2WApp(container: AppContainer) {
    T2WTheme {
        val factory = remember(container) { AppViewModelFactory(container) }
        val ready by container.session.ready.collectAsStateWithLifecycle()
        val tokens by container.session.tokens.collectAsStateWithLifecycle()

        Surface(
            modifier = Modifier.fillMaxSize(),
            color = MaterialTheme.colorScheme.background,
        ) {
            when {
                !ready -> LoadingView()
                tokens == null -> LoginScreen(factory = factory)
                else -> MainScreen(factory = factory)
            }
        }
    }
}

File: ios/T2W/App/T2WApp.swift
import SwiftUI

/// App entry + auth gate — the iOS analogue of the Android `T2WApp` composable.
/// It observes the session: while not ready it shows a spinner; `tokens == nil`
/// shows Login; otherwise the signed-in shell. Clearing the session anywhere
/// (logout, refresh failure) returns the user to Login automatically.
@main
struct T2WApp: App {
    @StateObject private var container = AppContainer()

    var body: some Scene {
        WindowGroup {
            RootView(container: container)
                .environmentObject(container.session)
        }
    }
}

private struct RootView: View {
    let container: AppContainer
    @EnvironmentObject private var session: SessionStore

    var body: some View {
        Group {
            if !session.isReady {
                ProgressView()
            } else if session.tokens == nil {
                // Build the login VM from the *real* container so a successful
                // login persists to the shared session and flips this gate.
                LoginView(viewModel: LoginViewModel(repository: container.authRepository))
            } else {
                MainTabsView(container: container)
            }
        }
    }
}

/// Signed-in shell. Mirrors the Android `MainScreen` bottom-nav with the same
/// four destinations (Home / Rides / Riders / Profile). This scaffold ships the
/// Rides tab wired to the client; the others are placeholders to fill in.
struct MainTabsView: View {
    let container: AppContainer

    var body: some View {
        TabView {
            Text("Tales on 2 Wheels")
                .tabItem { Label("Home", systemImage: "house") }

            RidesListView(repository: container.ridesRepository)
                .tabItem { Label("Rides", systemImage: "bicycle") }

            Text("Leaderboard")
                .tabItem { Label("Riders", systemImage: "trophy") }

            Text("Profile")
                .tabItem { Label("Profile", systemImage: "person") }
        }
    }
}
